File: engines/macos-swift/modules/rendering/BackgroundFramingGeometry.swift
import CoreGraphics
import Foundation
import Project

/// Resolution-independent background-stage geometry shared by preview and export rendering.
public struct BackgroundFramingGeometry: Equatable, Sendable {
    public let outputRect: CGRect
    public let cardRect: CGRect
    public let cornerRadius: CGFloat
    public let shadowOpacity: Float
    public let shadowRadius: CGFloat
    public let shadowOffset: CGSize

    public init?(
        renderSize: CGSize,
        orientedSourceSize: CGSize,
        settings: BackgroundFramingSettings
    ) {
        guard renderSize.isFiniteAndPositive, orientedSourceSize.isFiniteAndPositive else {
            return nil
        }

        let outputRect = CGRect(origin: .zero, size: renderSize)
        let shorterOutputDimension = min(renderSize.width, renderSize.height)
        let padding = settings.enabled ? CGFloat(settings.paddingFraction) * shorterOutputDimension : 0
        let availableRect = outputRect.insetBy(dx: padding, dy: padding)
        guard availableRect.width > 0, availableRect.height > 0 else { return nil }

        let scale = min(
            availableRect.width / orientedSourceSize.width,
            availableRect.height / orientedSourceSize.height
        )
        let cardSize = CGSize(
            width: orientedSourceSize.width * scale,
            height: orientedSourceSize.height * scale
        )
        let cardRect = CGRect(
            x: availableRect.midX - cardSize.width / 2,
            y: availableRect.midY - cardSize.height / 2,
            width: cardSize.width,
            height: cardSize.height
        )
        let shorterCardDimension = min(cardSize.width, cardSize.height)
        let shadowStrength = settings.enabled ? CGFloat(settings.shadowStrength) : 0

        self.outputRect = outputRect
        self.cardRect = cardRect
        cornerRadius = settings.enabled
            ? CGFloat(settings.cornerRadiusFraction) * shorterCardDimension
            : 0
        shadowOpacity = Float(0.30 * shadowStrength)
        shadowRadius = 0.035 * shorterOutputDimension * shadowStrength
        shadowOffset = CGSize(width: 0, height: 0.012 * shorterOutputDimension * shadowStrength)
    }
}

/// Explicit opaque sRGB color used by the deterministic background stage.
public struct BackgroundFramingColor: Equatable, Sendable {
    public let red: CGFloat
    public let green: CGFloat
    public let blue: CGFloat

    public init?(hex: String) {
        guard hex.count == 7, hex.first == "#", let value = UInt32(hex.dropFirst(), radix: 16) else {
            return nil
        }
        red = CGFloat((value >> 16) & 0xFF) / 255
        green = CGFloat((value >> 8) & 0xFF) / 255
        blue = CGFloat(value & 0xFF) / 255
    }

    public var cgColor: CGColor {
        let colorSpace = CGColorSpace(name: CGColorSpace.sRGB)!
        return CGColor(colorSpace: colorSpace, components: [red, green, blue, 1])!
    }
}

private extension CGSize {
    var isFiniteAndPositive: Bool {
        width.isFinite && height.isFinite && width > 0 && height > 0
    }
}

File: apps/desktop-electrobun/src/mainview/app/studio/domain/backgroundFramingGeometry.ts
import type { BackgroundFramingSettings } from "@guerillaglass/engine-contract/shared/valueObjects";

export type RenderSize = Readonly<{ width: number; height: number }>;
export type RenderRect = Readonly<{ x: number; y: number; width: number; height: number }>;

export type BackgroundFramingGeometry = Readonly<{
  outputRect: RenderRect;
  cardRect: RenderRect;
  cornerRadius: number;
  shadowOpacity: number;
  shadowBlurRadius: number;
  shadowOffsetX: number;
  shadowOffsetY: number;
}>;

export function computeBackgroundFramingGeometry(
  outputSize: RenderSize,
  sourceSize: RenderSize,
  settings: BackgroundFramingSettings,
): BackgroundFramingGeometry | null {
  if (!isFinitePositiveSize(outputSize) || !isFinitePositiveSize(sourceSize)) {
    return null;
  }

  const outputRect = { x: 0, y: 0, width: outputSize.width, height: outputSize.height };
  const shorterOutputDimension = Math.min(outputSize.width, outputSize.height);
  const padding = settings.enabled ? settings.paddingFraction * shorterOutputDimension : 0;
  const availableWidth = outputSize.width - 2 * padding;
  const availableHeight = outputSize.height - 2 * padding;
  if (availableWidth <= 0 || availableHeight <= 0) {
    return null;
  }

  const scale = Math.min(availableWidth / sourceSize.width, availableHeight / sourceSize.height);
  const cardWidth = sourceSize.width * scale;
  const cardHeight = sourceSize.height * scale;
  const cardRect = {
    x: (outputSize.width - cardWidth) / 2,
    y: (outputSize.height - cardHeight) / 2,
    width: cardWidth,
    height: cardHeight,
  };
  const shadowStrength = settings.enabled ? settings.shadowStrength : 0;

  return {
    outputRect,
    cardRect,
    cornerRadius: settings.enabled
      ? settings.cornerRadiusFraction * Math.min(cardWidth, cardHeight)
      : 0,
    shadowOpacity: 0.3 * shadowStrength,
    shadowBlurRadius: 0.035 * shorterOutputDimension * shadowStrength,
    shadowOffsetX: 0,
    shadowOffsetY: 0.012 * shorterOutputDimension * shadowStrength,
  };
}

function isFinitePositiveSize(size: RenderSize): boolean {
  return (
    Number.isFinite(size.width) && Number.isFinite(size.height) && size.width > 0 && size.height > 0
  );
}
